### PR TITLE
#155744144 Create login endpoint

### PR DESCRIPTION
**Pull request by muhozi**
#### What does this PR do?
Implement login endpoint
#### Description of Task to be completed?
Login and token issuing, User can login by providing email and password and get a token
#### How should this be manually tested?
After cloning the repo run `nosetest -v`
#### What are the relevant pivotal tracker stories?
https://www.pivotaltracker.com/story/show/155744144

### CODE REVIEW

File: tests/api_tests.py [hunk 1]
```diff
@@ -3,15 +3,16 @@
 """
 import unittest
 import uuid
-import json
 from api import app
 from api.models.user import user
 
+
 class main_tests(unittest.TestCase):
```
Comment (mkiterian) [inline tests/api_tests.py]: Review class naming conventions

File: tests/api_tests.py [hunk 2]
```diff
@@ -3,15 +3,16 @@
...
     """
         Main test
     """
     url_prefix = '/api/v1/'
+
     def setUp(self):
         """
             Set up tet data
```
Comment (mkiterian) [inline tests/api_tests.py]: Correct spelling

File: api/models/user.py [hunk 3]
```diff
@@ -1,18 +1,31 @@
+""" User Model """
+
+from werkzeug.security import generate_password_hash
 from api.models.store import store
+
+
 class user(object):
```
Comment (mkiterian) [inline api/models/user.py]: Review this: https://www.python.org/dev/peps/pep-0008/#class-names

File: tests/api_tests.py [hunk 4]
```diff
@@ -32,37 +33,65 @@ def setUp(self):
...
         })
         self.assertEqual(response.status_code, 200)
         self.assertIn(b'have been successfully registered', response.data)
+        # Test registration with incomplete data
+        incomplete_data_response = self.app.post(self.url_prefix + 'auth/register', data={
+            'names': 'dummy name',
+            'confirm_password': self.exist_user['confirm_password']
+        })
+        self.assertEqual(incomplete_data_response.status_code, 400)
+        self.assertIn(b'Please provide', incomplete_data_response.data)
```
Comment (mkiterian) [inline tests/api_tests.py]: What if an already registered user tries to register with the same credentials? Ensure all edge cases are covered in your tests

**Review by mkiterian — CHANGES_REQUESTED** [review]
Make sure your code conforms to PEP8 standards

File: api/api.py [hunk 5]
```diff
@@ -1,43 +1,105 @@
...
-        'names':request.form['names'],
-        'email':request.form['email'],
-        'password':generate_password_hash(request.form['password']),
+        'id': uuid.uuid4().hex,
+        'names': request.form['names'],
+        'email': request.form['email'],
+        'password': request.form['password'],
     }
-    if (user().user_exists(data['email'])):
+    if user().user_exists(data['email']):
```
Comment (mkiterian) [inline api/api.py]: Is there a test for this?

File: api/api.py [hunk 6]
```diff
@@ -1,43 +1,105 @@
...
         response.status_code = 400
         return response
     user().save(data)
     response = jsonify({
-        'status':'ok',
-        'message':"You have been successfully registered"
+        'status': 'ok',
+        'message': "You have been successfully registered"
     })
     response.status_code = 200
```
Comment (mkiterian) [inline api/api.py]: You are creating a new user, is 200 the correct status code?

File: api/api.py [hunk 7]
```diff
@@ -1,43 +1,105 @@
...
+    # Check if email exists in the store
+    logged_user = user().get_user(data['email'])
+    if logged_user:
+        # Check password
+        if check_password_hash(logged_user['password'], data['password']):
+            token = get_token(logged_user['id'])
+            user().add_token(token)
+            response = jsonify({
+                'status': 'ok',
+                'token': token,
```
Comment (mkiterian) [inline api/api.py]: Is the token being sent with the response data or is it set in the response headers?

File: api/inputs/inputs.py [hunk 8]
```diff
@@ -1,19 +1,32 @@
...
         self.error = ''
-    def validate(self,inputs):
-        if 'names' in inputs and 'email' in inputs and 'password' in inputs and 'confirm_password' in inputs:
+
+    def validate(self, inputs):
+        """ Register validation method """
+        if ('names' in inputs and
+                'email' in inputs and
+                'password' in inputs and
+                'confirm_password' in inputs):
```
Comment (mkiterian) [inline api/inputs/inputs.py]: Add validity checks for whether 'email' is an actual email, password is of correct length

File: tests/api_tests.py (hunk 2 above)
Comment (muhozi) [reply tests/api_tests.py]: Fixed

File: api/models/user.py (hunk 3 above)
Comment (muhozi) [reply api/models/user.py]: Fixed

File: tests/api_tests.py (hunk 4 above)
Comment (muhozi) [reply tests/api_tests.py]: Fixed

File: api/api.py (hunk 5 above)
Comment (muhozi) [reply api/api.py]: Fixed

File: api/api.py (hunk 7 above)
Comment (muhozi) [reply api/api.py]: Fixed
